lib/tbot/ca_rotation.go

```diff
@@ -0,0 +1,149 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tbot
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+// See https://github.com/gravitational/teleport/blob/1aa38f4bc56997ba13b26a1ef1b4da7a3a078930/lib/auth/rotate.go#L135
+// for server side details of how a CA rotation takes place.
+//
+// We can leverage the existing renewal system to fetch new certificates and
+// CAs.
+//
+// We need to force a renewal for the following transitions:
+// - Init -> Update Clients: So we can receive a set of certificates issued by
+//   the new CA, and trust both the old and new CA.
+// - Update Clients, Update Servers -> Rollback: So we can receive a set of
+//   certificates issued by the old CA, and stop trusting the new CA.
+// - Update Servers -> Standby: So we can stop trusting the old CA.
+
+func (b *Bot) caRotationLoop(ctx context.Context) error {
+	// TODO: Throw error if more than X consecutive errors in time window.
+	for ctx.Err() == nil {
+		err := b.watchCARotations(ctx)
+		if err != nil {
+			b.log.WithError(err).Warnf("Error ocurred whilst watching CA rotations, retrying...")
+		}
+	}
+
+	return nil
+}
+
+func (b *Bot) watchCARotations(ctx context.Context) error {
+	clusterName := b.ident().ClusterName
+	b.log.Debugf("Attempting to establish watch for CA events")
+	watcher, err := b.client().NewWatcher(ctx, types.Watch{
+		Kinds: []types.WatchKind{{
+			Kind: types.KindCertAuthority,
+			Filter: types.CertAuthorityFilter{
+				types.HostCA:     clusterName,
```

Comment:
As far as I can tell it's correct to filter by the cluster name on all of these, but I need someone more familiar than me with our CA to confirm this.